### PR TITLE
net-misc/streamlink-9999: add new needed dependency and missed PYTHON_REQ_USE

### PR DESCRIPTION
**Pull request by soredake**
Fixes https://bugs.gentoo.org/show_bug.cgi?id=601376, https://bugs.gentoo.org/show_bug.cgi?id=601374, https://bugs.gentoo.org/show_bug.cgi?id=601378 and https://bugs.gentoo.org/show_bug.cgi?id=601382

### CODE REVIEW

**Comment by Coacher** [comment]
metadata.xml for pycryptodome is missing.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+RESTRICT="test"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+gmp test"
+
+DEPEND="gmp? ( dev-libs/gmp:0 )
+				sys-devel/gcc
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Normally you shouldn't include deps from system set.
A comment why it's needed would be helpful.

Comment (soredake) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Gcc is listed as dep in https://github.com/Legrandin/pycryptodome/blob/master/INSTALL.rst#linux-fedora and used in this code https://github.com/Legrandin/pycryptodome/blob/master/setup.py#L139-L193

Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Re your first link: why GCC is explicitly required still unclear. Instructions for Ubuntu on the same page don't mention it.
Re your second link: no mentions of GCC on that page at all.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+				dev-python/setuptools"
+RDEPEND="${DEPEND}"
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Missing blocker on dev-python/pycrypto.
Both pycrypto and pycryptodome install `usr/lib/pythonX.X/site-packages/Crypto/` directory, where X.X is your python version.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild (hunk 1 above)
Comment (soredake) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: For ubuntu, build-essential pulls gcc, but well, I'll take away gcc

File: net-misc/streamlink/streamlink-9999.ebuild [hunk 3]
```diff
@@ -13,17 +14,19 @@ HOMEPAGE="https://streamlink.github.io/"
...
-LICENSE="BSD-2"
+LICENSE="BSD-2 MIT"
 SLOT="0"
 IUSE="doc test"
 
 RDEPEND="dev-python/pycrypto[${PYTHON_USEDEP}]
 	dev-python/requests[${PYTHON_USEDEP}]
 	virtual/python-futures[${PYTHON_USEDEP}]
 	virtual/python-singledispatch[${PYTHON_USEDEP}]
+	dev-python/pycryptodome[${PYTHON_USEDEP}]
```
Comment (Coacher) [inline net-misc/streamlink/streamlink-9999.ebuild]: Incorrect. pycrypto and pycryptodome are mutually exclusive.

File: dev-python/pycryptodome/metadata.xml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE pkgmetadata SYSTEM "http://www.gentoo.org/dtd/metadata.dtd">
+<pkgmetadata>
+	<maintainer type="person">
+		<email>fdsfgs@krutt.org</email>
+		<name>soredake</name>
+	</maintainer>
+	<maintainer type="project">
+			<email>proxy-maint@gentoo.org</email>
```
Comment (mgorny) [inline dev-python/pycryptodome/metadata.xml]: Misindent, looks like.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# Copyright 1999-2016 Gentoo Foundation
```
Comment (mgorny) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Adding the new package really needs to go into a separate commit (preceding the one adding the dep).

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+# $Id$
+
+EAPI=6
+PYTHON_COMPAT=( python{2_7,3_4,3_5} )
+# needed to fix the: No such file or directory: 'build/test1.c'
+DISTUTILS_IN_SOURCE_BUILD="1"
+
+inherit distutils-r1
+
+DESCRIPTION="Self-contained Python package of low-level cryptographic primitives."
```
Comment (mgorny) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: This is not a proper sentence, so no dot (.)

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+RESTRICT="test"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+gmp test"
+
+DEPEND="!dev-python/pycrypto
+				gmp? ( dev-libs/gmp:0 )
```
Comment (mgorny) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Another terrible misindent ;-). Use one tab for one indentation level. I feel like you're using an editor with two-space wide tab and trying to align stuff.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+DESCRIPTION="Self-contained Python package of low-level cryptographic primitives."
+HOMEPAGE="https://github.com/Legrandin/pycryptodome"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+RESTRICT="test"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="+gmp test"
```
Comment (mgorny) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Err, I don't see USE=gmp controlling anything. Is that a pure runtime dep kind of thing?

File: net-misc/streamlink/streamlink-0.1.0.ebuild [hunk 9]
```diff
@@ -4,6 +4,7 @@
 
 EAPI=6
 
+PYTHON_REQ_USE='xml,threads'
```
Comment (mgorny) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: It is recommended to `(+)` those flags, for compatibility with other impls. See: https://wiki.gentoo.org/wiki/Project:Python/Implementation_USE_flags

File: net-misc/streamlink/streamlink-0.1.0.ebuild [hunk 10]
```diff
@@ -13,17 +14,17 @@ HOMEPAGE="https://streamlink.github.io/"
...
-RDEPEND="dev-python/pycrypto[${PYTHON_USEDEP}]
-	dev-python/requests[${PYTHON_USEDEP}]
+RDEPEND="dev-python/requests[${PYTHON_USEDEP}]
 	virtual/python-futures[${PYTHON_USEDEP}]
 	virtual/python-singledispatch[${PYTHON_USEDEP}]
 	media-video/rtmpdump"
 DEPEND="dev-python/setuptools[${PYTHON_USEDEP}]
-	doc? ( dev-python/sphinx[${PYTHON_USEDEP}] )
+	doc? ( dev-python/sphinx[${PYTHON_USEDEP}] 
+				dev-python/docutils[${PYTHON_USEDEP}] )
```
Comment (mgorny) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: Misindent: here two tabs would be perfect.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild (hunk 7 above)
Comment (soredake) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: from my side it looked pretty normal https://a.uguu.se/s77cejJvxn2Z_p1dYCBZ%28e%2CZ%3A2TP%2ABG%7D.png

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild (hunk 5 above)
Comment (soredake) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Ok, I'll split after all other problems are resolved.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+
+DEPEND="!dev-python/pycrypto
+	dev-libs/gmp:0
+	dev-python/setuptools"
+RDEPEND="${DEPEND}"
```
Comment (SoapGentoo) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: can you fine tune these? i.e.
```
RDEPEND="!dev-python/pycrypto
	dev-libs/gmp:0="
DEPEND="${RDEPEND}
	dev-python/setuptools"
```
Notice the subslot operator on gmp. Unless you have very good reasons, setuptools should not be in `DEPEND`

File: net-misc/streamlink/streamlink-9999.ebuild [hunk 12]
```diff
@@ -4,6 +4,7 @@
 
 EAPI=6
 
+PYTHON_REQ_USE='xml(+),threads(+)'
```
Comment (SoapGentoo) [inline net-misc/streamlink/streamlink-9999.ebuild]: it is is customary to add `PYTHON_REQ_USE` always after `PYTHON_COMPAT`

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+RESTRICT="test"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+
+DEPEND="!dev-python/pycrypto
+	dev-libs/gmp:0
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: dev-libs/gmp should be in neither DEPEND, nor RDEPEND: https://github.com/Legrandin/pycryptodome/blob/master/Changelog.rst#31-15-march-2015 GMP library is loaded via cffi/ctypes at runtime when present, but pycryptodome can work without it. In such cases you usually just mention dev-libs/gmp dependency in pkg_postinst() via elog/optfeature.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+
+RDEPEND="!dev-python/pycrypto"
+DEPEND="${RDEPEND}
+	dev-python/setuptools"
+
+pkg_postinst() {
+	optfeature "For faster public key operations" dev-libs/gmp
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: optfeature requires eutils eclass.

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+	}
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: Misindent

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+HOMEPAGE="https://github.com/Legrandin/pycryptodome"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+RESTRICT="test"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+
+RDEPEND="!dev-python/pycrypto"
```
Comment (Coacher) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: A bit more subtle approach is required here.
The number of packages that want pycryptodome is one: streamlink
The number of packages that want pycrypto is huge: https://qa-reports.gentoo.org/output/genrdeps/rindex/dev-python/pycrypto

Thus you are denying users from having streamlink and anything from the list above coinstalled.
One way to avoid this is to update all the packages from the list above to allow either pycrypto or pycryptodome.
But maybe there is a better way.

**Comment by Whissi** [comment]
@soredake: *ping*

Are you still interested in proxy-maintaining these packages? Would be nice to get dev-python/pycryptodome/pycryptodome into tree (see https://bugs.gentoo.org/show_bug.cgi?id=602022) to last-ride dev-python/pycrypto package (dead upstream).

File: dev-python/pycryptodome/pycryptodome-3.4.3.ebuild (hunk 16 above)
Comment (saboya) [inline dev-python/pycryptodome/pycryptodome-3.4.3.ebuild]: This patch will make streamlink use Cryptodome instead of Crypto:

```diff
--- streamlink-0.2.0/setup.py
+++ streamlink-0.2.0/setup.py
@@ -38,7 +38,7 @@
     deps.append("requests>=1.0,!=2.12.0,!=2.12.1,<3.0")
 
 # this version of pycryptodome is known to work and has a Windows wheel for py2.7, py3.3-3.5
-deps.append("pycryptodome==3.4.3")
+deps.append("pycryptodomex==3.4.3")
 
 # When we build an egg for the Win32 bootstrap we don't want dependency
 # information built into it.
--- streamlink-0.2.0/src/streamlink/plugins/beattv.py
+++ streamlink-0.2.0/src/streamlink/plugins/beattv.py
@@ -3,7 +3,7 @@
 from collections import namedtuple
 from io import BytesIO
 try:
-    from Crypto.Cipher import AES
+    from Cryptodome.Cipher import AES
     CAN_DECRYPT = True
 except ImportError:
     CAN_DECRYPT = False
--- streamlink-0.2.0/src/streamlink/stream/hls.py
+++ streamlink-0.2.0/src/streamlink/stream/hls.py
@@ -1,7 +1,7 @@
 from collections import defaultdict, namedtuple
 
 try:
-    from Crypto.Cipher import AES
+    from Cryptodome.Cipher import AES
     import struct
 
     def num_to_iv(n):```

It will allow Crypto and Cryptodome to co-exist, since Cryptodome will be installed under Cryptodomex. However, I think the correct approach would be a virtual/python-crypto package to handle the migration, since pycrypto is abandoned and cryptodome is a drop-in replacement.

**Comment by soredake** [comment]
@Whissi i'm still interested, but this
>A bit more subtle approach is required here.
The number of packages that want pycryptodome is one: streamlink
The number of packages that want pycrypto is huge: https://qa-reports.gentoo.org/output/genrdeps/rindex/dev-python/pycrypto

is still a problem, before adding pycryptodome to tree we need to test all this packages to confirm that they are working with new library.

**Comment by Whissi** [comment]
Good to hear that you are still interested.

> is still a problem, before adding pycryptodome to tree we need to test all this packages to confirm that they are working with new library.

Because *pycryptodome* can be used as drop-in replacement for *pycrypto* and is therefore causing file collisions, right?

I guess we have to create a virtual package for *pycrypto*. But I hope @gentoo/python will assist you here.

**Comment by saboya** [comment]
dev-python/backports-shutil_which is also needed. PR #3687

**Comment by saboya** [comment]
dev-python/iso3166 is needed since 0.3.1. PR #3887 
dev-ptyhon/iso_639 is needed since 0.3.1. PR #3886

**Comment by Coacher** [comment]
9999 also has pycountry support as a fallback. Should be easier to backport the required commit than support extra python packages.

**Comment by Techwolf** [comment]
dev-python/backports-shutil_which only has python2_7. That conflicts with python{2_7,3_4,3_5} in the streamlink ebuild.

The error I get is:
```
emerge: there are no ebuilds built with USE flags to satisfy "dev-python/backports-shutil_which[python_targets_python2_7(-)?,python_targets_python3_4(-)?,python_targets_python3_5(-)?,-python_single_target_python2_7(-),-python_single_target_python3_4(-),-python_single_target_python3_5(-)]".
!!! One of the following packages is required to complete your request:
- net-misc/streamlink-0.3.1::Techwolf (Change USE: -python_targets_python3_4)
```

File: net-misc/streamlink/streamlink-0.3.1.ebuild [hunk 17]
```diff
@@ -1,30 +1,40 @@
...
+LICENSE="BSD-2 MIT"
 SLOT="0"
 IUSE="doc test"
 
-RDEPEND="dev-python/pycrypto[${PYTHON_USEDEP}]
+RDEPEND="dev-python/pycryptodome[${PYTHON_USEDEP}]
 	dev-python/requests[${PYTHON_USEDEP}]
 	virtual/python-futures[${PYTHON_USEDEP}]
 	virtual/python-singledispatch[${PYTHON_USEDEP}]
+	dev-python/backports-shutil_which[${PYTHON_USEDEP}]
```
Comment (saboya) [inline net-misc/streamlink/streamlink-0.3.1.ebuild]: Should be changed to

```diff
+	dev-python/backports-shutil_which[$(python_gen_usedep 'python2*')]
```

File: net-misc/streamlink/streamlink-0.3.1.ebuild [hunk 18]
```diff
@@ -1,30 +1,40 @@
...
+	dev-python/backports-shutil_get_terminal_size[${PYTHON_USEDEP}]
```
Comment (saboya) [inline net-misc/streamlink/streamlink-0.3.1.ebuild]: Should be changed to

```diff
+	dev-python/backports-shutil_get_terminal_size[$(python_gen_usedep 'python2*')]
```

**Comment by saboya** [comment]
@Techwolf It's because those packages are backports from python3, and the package shouldn't depend on them. The changes I proposed above fix those problems.

**Comment by Techwolf** [comment]
@saboya The changes did it. It now builds for 2.7 and 3.4 on my system.

**Comment by mgorny** [comment]
@soredake, ping, are you still intereted in this? pycryptodome is now in ::gentoo, so you can move on if that was the problem.

**Comment by soredake** [comment]
@mgorny still, I've updated the commit, i think we can merge it now.

File: net-misc/streamlink/streamlink-0.3.2.ebuild [hunk 19]
```diff
@@ -1,29 +1,35 @@
...
+	dev-python/backports-shutil_which[$(python_gen_usedep 'python2*')]
+	dev-python/backports-shutil_get_terminal_size[$(python_gen_usedep 'python2*')]
+	dev-python/pycountry[${PYTHON_USEDEP}]
 	media-video/rtmpdump"
 DEPEND="dev-python/setuptools[${PYTHON_USEDEP}]
-	doc? ( dev-python/sphinx[${PYTHON_USEDEP}] )
-	test? ( ${RDEPEND} )"
+	doc? ( dev-python/sphinx[${PYTHON_USEDEP}]
+		dev-python/docutils[${PYTHON_USEDEP}] )
+	test? ( dev-python/mock
```
Comment (mgorny) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: PYTHON_USEDEP missing here.

Comment (Coacher) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: Since 0.3.0 you only need dev-python/mock with Python-2.

File: net-misc/streamlink/streamlink-0.3.2.ebuild [hunk 20]
```diff
@@ -1,29 +1,35 @@
...
 IUSE="doc test"
 
-RDEPEND="dev-python/pycrypto[${PYTHON_USEDEP}]
+RDEPEND="dev-python/pycryptodome[${PYTHON_USEDEP}]
 	dev-python/requests[${PYTHON_USEDEP}]
 	virtual/python-futures[${PYTHON_USEDEP}]
 	virtual/python-singledispatch[${PYTHON_USEDEP}]
+	dev-python/backports-shutil_which[$(python_gen_usedep 'python2*')]
+	dev-python/backports-shutil_get_terminal_size[$(python_gen_usedep 'python2*')]
+	dev-python/pycountry[${PYTHON_USEDEP}]
```
Comment (Coacher) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: Not enough. You need to prevent those iso* packages from creeping into egg like this:
```
python_prepare_all() {
	# Avoid iso-639, iso3166 dependencies since we use pycountry.
	export STREAMLINK_USE_PYCOUNTRY=1

	distutils-r1_python_prepare_all
}
```

File: net-misc/streamlink/streamlink-0.3.2.ebuild [hunk 21]
```diff
@@ -1,29 +1,35 @@
...
 	media-video/rtmpdump"
```
Comment (Coacher) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: Missing virtual/ffmpeg rdep. See src/streamlink/stream/ffmpegmux.py

File: net-misc/streamlink/streamlink-0.3.2.ebuild (hunk 20 above)
Comment (mgorny) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: Don't declare `prepare` for that. Just use `python_configure_all()` which has the advantage of not having a default to call.

Comment (Coacher) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: Yeah, even better.

File: net-misc/streamlink/streamlink-0.1.0.ebuild [hunk 22]
```diff
@@ -1,39 +0,0 @@
-# Copyright 1999-2016 Gentoo Foundation
```
Comment (Coacher) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: First you remove the last keyworded version from tree leaving a dangling Manifest entry behind. Then proceed with the bump to reintroduce a keyworded ebuild and update Manifest.

Even though this will be pushed in one go, it would be nice to avoid tree breakage at any given point and first do bump and only then remove previous version. Coud you please reorder your commits and adjust Manifest handling.

Comment (soredake) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: Sorry! Reordered.

Comment (Coacher) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: Thanks. One more thing. Why did you drop ~arm and ~x86 KEYWORDS?

Comment (soredake) [inline net-misc/streamlink/streamlink-0.1.0.ebuild]: Added them back.

**Comment by Coacher** [comment]
LGTM.
@soredake, thank you very much for all the work on this one.

File: net-misc/streamlink/streamlink-0.3.2.ebuild [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+PYTHON_COMPAT=( python{2_7,3_4,3_5} )
+PYTHON_REQ_USE='xml(+),threads(+)'
+
+inherit distutils-r1
+
+DESCRIPTION="CLI for extracting streams from websites to a video player of your choice"
+HOMEPAGE="https://streamlink.github.io/"
+SRC_URI="https://github.com/${PN}/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
+
+KEYWORDS="~amd64 ~arm ~x86"
```
Comment (mgorny) [inline net-misc/streamlink/streamlink-0.3.2.ebuild]: You have unkeyworded deps on ~arm, so you need to drop that kw. My suggestion would be to drop it here, leave the older ebuild alone if it still works (i.e. keep it) and we'll make a big keywording for pycryptodome and revdeps at some point.

**Comment by mgorny** [comment]
@soredake, please let me know whether there is any point in keeping the old version, i.e. if it works at all. Please also close the bugs if you can or ping us if you can't.

**Comment by soredake** [comment]
@mgorny as long as it works we can keep it.

**Comment by flyser** [comment]
Since STREAMLINK_USE_PYCOUNTRY does not have any effect in 0.3.2, you need to cherry-pick https://github.com/streamlink/streamlink/pull/585 to

**Comment by soredake** [comment]
@flyser https://github.com/gentoo/gentoo/pull/4130
